Use regexp instead of grep for excluding files

karelbilek opened the pull request:
It's more clear and should work the same

karelbilek committed:
Use regexp instead of grep for excluding files

It's more clear and should work the same

alex-parra commented on internal/main.go:
```diff
@@ -33,8 +35,26 @@ func RunTests(opts RunTestsOpts) {
+		testPkgs = []string{}
+		excludeRegexs := make([]*regexp.Regexp, 0, len(opts.Excludes))
+		for _, exclude := range opts.Excludes {
+			regex, err := regexp.Compile("^" + exclude + "$")
```
The existing code doesn't add `$` at the end. 
It is so that each entry is treated as a prefix while also allowing the user to add the `$` where needed.

karelbilek replied:
ok

karelbilek replied:
Donne

alex-parra commented on internal/main.go:
```diff
@@ -33,8 +35,26 @@ func RunTests(opts RunTestsOpts) {
+		}
+		for _, pkg := range allPkgs {
+			isIncluded := true
+			for _, regex := range excludeRegexs {
```
Wouldn't it be more performant to make a single regex and thus not have the nested loops.
Thinking `^first/exclude|^second/exclude|^etc/etc`? 
It also means a single regex comparison per tested package.

karelbilek replied:
I'm just not sure if it doesn't have some edge-case, like if the regexes are not properly formed and the | will somehow do some nonsense

karelbilek replied:
I think the performance difference will be negligent (especially we spend most time on waiting on `go test` later anyway)

alex-parra reviewed (changes requested):
Thanks!
Agree, regex instead of grep should be better since it's all done in go instead of crossing to the shell. 
Left a couple of comments.

alex-parra commented on internal/main.go:
```diff
@@ -33,8 +35,26 @@ func RunTests(opts RunTestsOpts) {
 	if len(opts.Excludes) > 0 {
-		testPkgsStr := shCmd("grep", shArgs{"-Ev", getExcludePattern(opts.Excludes)}, allPkgsStr)
-		testPkgs = splitLines(testPkgsStr)
+		testPkgs = []string{}
```
Also, if we move this whole block to a function it's easy to put it under a test.

karelbilek replied:
Done

karelbilek committed:
Moved to separate file/function, added tests

I have also removed no longer use generic function. We can return it back if we need it I guess

karelbilek commented:
I moved the func now. See the other commit

karelbilek committed:
Fix test regexp

alex-parra reviewed (approved):
Nice! Thanks
